persistence/upgrades/nhibernate-7to8.md

```diff
@@ -0,0 +1,49 @@
+
+Processes A and B read value `5` as the current value of the counter and both processes increment the value in memory. Then both processes update the database. Without the optimistic concurrency check the outcome is `6` which is incorrect because we expected to increment the value twice.
+
+With optimistic concurrency check the processes would append `where counter = 5` clause to the update statement to ensure that the value being updated has not been modified. One of the processes would succeed and the other would fail (no rows updated) and would have to re-try, this time reading value `6` as the current counter.
+
+## Accessing DateTime columns using DateTime2 type
+
+By default, NHibernate persistence uses all Saga fields for optimistic concurrency control so the generated update statement contains a `where` clause with all previously read Saga property values.
+
+When reading `DateTime` column values using `DateTime2(7)` type the value is read without any precision loss and stored in the .NET `DateTime` type. Later, when building the update command NHibernate passes the `DateTime2(7)` type as the command parameter type which causes ADO.NET to send a `DateTime2(7)` value to the database rather than `DateTime` value. Because these are different types, the comparison fails resulting in `ObjectStaleException` being thrown by NHibernate.
+
+## Upgrade paths
+
+Because, as mentioned, NHibernate 5 defaults to `DateTime2(7)` type, all new saga tables will used that new type for storing .NET `DateTime` values. These new sagas will work correctly with Version 8.1 of NHibernate persistence.
+
+All the existing sagas that use `DateTime` type in their tables stop working with Version 8.1 of NHibernate persistence due to the optimistic concurrency problem manifesting in `ObjectStaleException` being thrown. There are three upgrade paths to make these existing sagas work.
+
+### Alter schema
+
+One way to solve the problem is to alter the schema of existing saga tables to use the new `DateTime2(7)` data type. Such alteration can be easily done in-place without any data/precision loss.
+
+### Use explicit version field
+
+NHibernate persistence has an option to use [explicit version column](/persistence/nhibernate/saga-concurrency.md#explicit-version) for optimistic concurrency checks. This is usually a better option because the resulting update command is much simpler. Enabling the explicit versioning requires changing the sagas code and altering the saga table schema (adding a column). 
+
+### Force backward compatibility
+
+If altering the schema is not an option, NHibernate offers a configuration switch to fall back to using the classic `DateTime` data type
+
+snippet: NHibernateLegacyDateTimeUpgrade7To8
+
+The downside of this approach is that this is a global setting and affects all tables mapped using that NHibernate configuration object.
```

Comment:
@danielmarbach I realized this only when I wrote it. It means that users can't really mix `DT` and `DT2` types. This might be a problem if the use NServiceBus-managed session to access their data and they use `DT2` in their own entity tables.

Reply:
Have you seen this? https://github.com/nhibernate/nhibernate-core/blob/ac39173567b31bcfad475ca32687c9faf0d37f87/src/NHibernate.Test/TypesTest/DateTimeTypeFixture.cs#L89-L100

Could users implement their own session factory and then manipulate local configuration?

Reply:
LoL. That's evil. I'd rather not document that.